added MobSF_Service analyzer

spoiicy opened the pull request:
Closes #2496 

# Description

Added MobSF_Service analyzer

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

# Checklist

- [x] I have read and understood the rules about [how to Contribute](https://intelowlproject.github.io/docs/IntelOwl/contribute/) to this project
- [x] The pull request is for the branch `develop`
- [x] A new plugin (analyzer, connector, visualizer, playbook, pivot or ingestor) was added or changed, in which case:
    - [x] I strictly followed the documentation ["How to create a Plugin"](https://intelowlproject.github.io/docs/IntelOwl/contribute/#how-to-add-a-new-plugin)
    - [ ] [Usage](https://github.com/intelowlproject/docs/blob/main/docs/IntelOwl/usage.md) file was updated.
    - [ ] [Advanced-Usage](https://github.com/intelowlproject/docs/blob/main/docs/IntelOwl/advanced_usage.md) was updated (in case the plugin provides additional optional configuration).
    - [x] I have dumped the configuration from Django Admin using the `dumpplugin` command and added it in the project as a data migration. (["How to share a plugin with the community"](https://intelowlproject.github.io/docs/IntelOwl/contribute/#how-to-share-your-plugin-with-the-community))
    - [ ] If a File analyzer was added and it supports a mimetype which is not already supported, you added a sample of that type inside the archive `test_files.zip` and you added the default tests for that mimetype in [test_classes.py](https://github.com/intelowlproject/IntelOwl/blob/master/tests/api_app/analyzers_manager/test_classes.py).
    - [ ] If you created a new analyzer and it is free (does not require any API key), please add it in the `FREE_TO_USE_ANALYZERS` playbook by following [this guide](https://intelowlproject.github.io/docs/IntelOwl/contribute/#how-to-modify-a-plugin).
    - [ ] Check if it could make sense to add that analyzer/connector to other [freely available playbooks](https://intelowlproject.github.io/docs/IntelOwl/usage/#list-of-pre-built-playbooks).
    - [x] I have provided the resulting raw JSON of a finished analysis and a screenshot of the results.
    - [ ] If the plugin interacts with an external service, I have created an attribute called precisely `url` that contains this information. This is required for Health Checks. 
    - [x] If the plugin requires mocked testing, `_monkeypatch()` was used in its class to apply the necessary decorators.
    - [x] I have added that raw JSON sample to the `MockUpResponse` of the `_monkeypatch()` method. This serves us to provide a valid sample for testing.
- [ ] If external libraries/packages with restrictive licenses were used, they were added in the [Legal Notice](https://github.com/certego/IntelOwl/blob/master/.github/legal_notice.md) section.
- [x] Linters (`Black`, `Flake`, `Isort`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowlproject.github.io/docs/IntelOwl/contribute/#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [ ] I have added tests for the feature/bug I solved (see `tests` folder). All the tests (new and old ones) gave 0 errors.
- [ ] If changes were made to an existing model/serializer/view, the docs were updated and regenerated (check [CONTRIBUTE.md](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Contribute.md)).
- [ ] If the GUI has been modified:
    - [ ] I have a provided a screenshot of the result in the PR.
    - [ ] I have created new frontend tests for the new component or updated existing ones.
- [x] After you had submitted the PR, if `DeepSource`, `Django Doctors` or other third-party linters have triggered any alerts during the CI checks, I have solved those alerts.

### Important Rules
- If you miss to compile the Checklist properly, your PR won't be reviewed by the maintainers.
- Everytime you make changes to the PR and you think the work is done, you should explicitly ask for a review. After being reviewed and received a "change request", you should explicitly ask for a review again once you have made the requested changes.

spoiicy commented:
<img width="1470" alt="Screenshot 2024-11-29 at 3 41 34 PM" src="https://github.com/user-attachments/assets/e411fa2c-b0dd-492a-84c0-21722907690b">
Attached is the screenshot for the MobSF_Service report and below is the full JSON report containing results for static & dynamic analysis. 
[job#38_report.json](https://github.com/user-attachments/files/17957663/job.38_report.json)

spoiicy commented:
@mlodic Could you please re-run the workflow since the backend-test CI has failed for some reason. It was passing in the previous commit [ffc968b](https://github.com/intelowlproject/IntelOwl/pull/2584/commits/ffc968b1849661101c98613057d84b673a999865)

mlodic commented on api_app/analyzers_manager/migrations/0139_analyzer_config_mobsf_service.py:
```diff
@@ -0,0 +1,150 @@
+    "docker_based": False,
+    "maximum_tlp": "RED",
+    "observable_supported": [],
+    "supported_filetypes": [
+        "application/vnd.android.package-archive",
+        "application/zip",
+    ],
```
you can use this list for android:

application/java-archive
application/vnd.android.package-archive
application/x-dex
application/zip

mlodic commented on api_app/analyzers_manager/migrations/0139_analyzer_config_mobsf_service.py:
```diff
@@ -0,0 +1,150 @@
+    "health_check_status": True,
+    "type": "file",
+    "docker_based": False,
+    "maximum_tlp": "RED",
```
Even if you should have control of the MOBSF Instance, it is still an external service so by definition this should be AMBER

mlodic commented on api_app/analyzers_manager/file_analyzers/mobsf_service.py:
```diff
@@ -0,0 +1,128 @@
+            headers=headers,
+        )
+        upload_response.raise_for_status()
+        scan_hash = upload_response.json()["hash"]
```
add a log here please

mlodic commented on api_app/analyzers_manager/file_analyzers/mobsf_service.py:
```diff
@@ -0,0 +1,128 @@
+    def run(self):
+        headers = {"X-Mobsf-Api-Key": self._mobsf_api_key}
+        binary = self.read_file_bytes()
+
```
add a log here please

mlodic commented on api_app/analyzers_manager/file_analyzers/mobsf_service.py:
```diff
@@ -0,0 +1,128 @@
+        pass
+
+    def static_analysis(self, scan_hash, headers):
+
```
add a log here please

mlodic commented on api_app/analyzers_manager/file_analyzers/mobsf_service.py:
```diff
@@ -0,0 +1,128 @@
+        data = {"hash": scan_hash}
+        scan_response = requests.post(url=scan_url, data=data, headers=headers)
+        scan_response.raise_for_status()
+
```
add a log here please

mlodic commented on api_app/analyzers_manager/file_analyzers/mobsf_service.py:
```diff
@@ -0,0 +1,128 @@
+        report_response = requests.post(url=report_url, data=data, headers=headers)
+        report_response.raise_for_status()
+        return report_response.json()
+
```
add a log here please

mlodic commented on api_app/analyzers_manager/file_analyzers/mobsf_service.py:
```diff
@@ -0,0 +1,128 @@
+        mobsfy_runtime_response = requests.post(
+            url=self.mobsf_host + "/api/v1/android/mobsfy",
+            headers=headers,
+            data={"identifier": "127.0.0.1:6555"},
```
this should be a parameter cause it can change between environments

mlodic commented on api_app/analyzers_manager/file_analyzers/mobsf_service.py:
```diff
@@ -0,0 +1,128 @@
+class MobSF_Service(FileAnalyzer):
+    mobsf_host: str
+    _mobsf_api_key: str
+
```
I would add a parameter to enable or not dynamic analysis. Static is the default. Dynamic could be opted out to save time and computation.

mlodic commented on api_app/analyzers_manager/file_analyzers/mobsf_service.py:
```diff
@@ -0,0 +1,128 @@
+
+    def static_analysis(self, scan_hash, headers):
+
+        scan_url = self.mobsf_host + "/api/v1/scan"
+        data = {"hash": scan_hash}
+        scan_response = requests.post(url=scan_url, data=data, headers=headers)
+        scan_response.raise_for_status()
+
+        report_url = self.mobsf_host + "/api/v1/report_json"
+        report_response = requests.post(url=report_url, data=data, headers=headers)
+        report_response.raise_for_status()
+        return report_response.json()
+
```
about the static analysis I am worried about execution time. By default, `requests` does not set any timeout. This is potentially dangerous in IntelOwl because it could lead to jobs that never end, creating bottlenecks in the overall application.
IntelOwl cannot rely on external tools. This is the reason why we need to explicity set timeouts to all the `requests` calls that you made in this analyzer. Which value? What your experience with this tool tells you? I think that static analysis could last several seconds. 30 seconds as default could make sense?
Plus, this value should be a parameter of the analyzer and customized so the users are not bound to it.

spoiicy replied:
Though i've not used mobsf extensively but the apps i've tested did not take more than 30 seconds for static analysis. I understand it will be better if we set a parameter for this so that the end-user can modify it according to their needs.

But I wan't to understand, do we need to set this timeout because when I was implementing this analyzer, it was failing at the "dynamic analysis" stage with an error "SoftTimeLimitExceeded". So I had increased the Soft-time limit in the analyzer config.

I do understand that setting a default timeout would be better for error handling but wouldn't just configuring the soft-time limit take care of everything. Let me know what you think!

mlodic replied:
> Even in the worst-case scenario, where our mobsf doesn't respond to the request and the request keeps hanging, wouldn't celery's Soft time limit take care of this and would end the job after specific period of time?

Yes but that would be the _last_ resort and should be treated as an unmanaged exception. It's like creating a generic Exception in python code. Ok, you get when something bad happens but you did not manage it properly cause you don't know what exactly was happening. Pretty much the same thing here. Soft time limit should be treated as the emergency exception.

About adding the timeout, we already did that for other similar services if you check other analyzers like [this](https://github.com/intelowlproject/IntelOwl/blob/master/api_app/analyzers_manager/observable_analyzers/circl_pdns.py#L39). High soft time limits and customizable timeouts, with low timeout values by default.

Regarding the values that you proposed, I think that even more than 400 could be set, maybe 1000 to handle the worst cases scenarios.

spoiicy replied:
Completely get your point about the _unmanaged exception_. I'll provide an option of request timeout parameter with a reasonable default value which can be modified as per user's need. 

I just want to understand one last thing, let's say a mobsf job has been triggered where default soft-time limit is 60 seconds but now the user has set the request timeout value as 400 seconds. Would this 400 seconds timeout be respected or the job will fail with error "SoftTimeLimitExceeded". 

Because from my limited understanding of the Intelowl, if the alloted time for a job to run is set to 60 seconds, the job will end even if we set the request timeout higher than the soft-time limit. So along with setting the request timeout, we'll have to configure the soft-time limit to a decent time period so that the job succeeds.

Let me know if my understanding is correct/incorrect?

mlodic replied:
> I just want to understand one last thing, let's say a mobsf job has been triggered where default soft-time limit is 60 seconds but now the user has set the request timeout value as 400 seconds. Would this 400 seconds timeout be respected or the job will fail with error "SoftTimeLimitExceeded".

the job would fail with SoftTimeLimitExceeded.

> we'll have to configure the soft-time limit to a decent time period so that the job succeeds.

Exactly. In this way soft time limits exception are limited to unwanted exceptions. I understand that the timeout can be changed by the user in a way that the soft time limit could always be reached but that's something we can ignore to avoid over-engineering this

mlodic commented on api_app/analyzers_manager/file_analyzers/mobsf_service.py:
```diff
@@ -0,0 +1,128 @@
+        return report_response.json()
+
+    def dynamic_analysis(self, scan_hash, headers):
+
```
add a log here please

mlodic commented on api_app/analyzers_manager/file_analyzers/mobsf_service.py:
```diff
@@ -0,0 +1,128 @@
+            data={"identifier": "127.0.0.1:6555"},
+        )
+        mobsfy_runtime_response.raise_for_status()
+
```
add a log here please

mlodic commented on api_app/analyzers_manager/file_analyzers/mobsf_service.py:
```diff
@@ -0,0 +1,128 @@
+            data={"hash": scan_hash},
+        )
+        start_dynamic_analysis_response.raise_for_status()
+
```
add a log here please

mlodic commented on api_app/analyzers_manager/file_analyzers/mobsf_service.py:
```diff
@@ -0,0 +1,128 @@
+            data={"hash": scan_hash},
+        )
+        tls_tests_response.raise_for_status()
+
```
add a log here please

mlodic commented on api_app/analyzers_manager/file_analyzers/mobsf_service.py:
```diff
@@ -0,0 +1,128 @@
+            },
+        )
+        start_frida_instrumentation_response.raise_for_status()
+
```
add a log here please

mlodic commented on api_app/analyzers_manager/file_analyzers/mobsf_service.py:
```diff
@@ -0,0 +1,128 @@
+            data={"hash": scan_hash},
+        )
+        get_runtime_dependency_response.raise_for_status()
+
```
add a log here please

mlodic commented on api_app/analyzers_manager/file_analyzers/mobsf_service.py:
```diff
@@ -0,0 +1,128 @@
+            data={"hash": scan_hash},
+        )
+        stop_dynamic_analysis.raise_for_status()
+
```
add a log here please

mlodic commented on api_app/analyzers_manager/file_analyzers/mobsf_service.py:
```diff
@@ -0,0 +1,128 @@
+            data={"hash": scan_hash},
+        )
+        dynamic_analysis_report.raise_for_status()
+
```
add a log here please

mlodic commented on api_app/analyzers_manager/file_analyzers/mobsf_service.py:
```diff
@@ -0,0 +1,128 @@
+        get_runtime_dependency_response.raise_for_status()
+
+        stop_dynamic_analysis = requests.post(
+            url=self.mobsf_host + "/api/v1/dynamic/stop_analysis",
```
as you suggested, dynamic analysis time is debatable. This is the reason why I think we should add here a sleep and that this should be a parameter of the analyzer. Which default? 20-30 seconds? What do you think about?

spoiicy replied:
As I commented earlier, in static analysis review comment, wouldn't just configuring the soft-time limit do the trick. Because if I keep it at default value, "SoftTimeLimitExceeded" error is thrown and my job fails. I had to set it to 400 in order for my job to be successful. Let me know what you think!

mlodic replied:
see previous comment

mlodic replied:
I noticed that no sleep was added so basically the analysis has a really brief duration time, just the time needed to execute all the steps between the calls "start" and "stop". As mentioned, I would add a sleep between the "start" and the later calls, to allow time for the stuff to start properly and the code loaded and maybe to allow the malware to do some nasty things too.

mlodic commented on api_app/analyzers_manager/file_analyzers/mobsf_service.py:
```diff
@@ -0,0 +1,128 @@
+            headers=headers,
+            data={
+                "hash": scan_hash,
+                "default_hooks": "api_monitor,ssl_pinning_bypass,root_bypass,debugger_check_bypass",
+                "auxiliary_hooks": "",
+                "frida_code": "",
```
these 3 values should be parameters of the analyzer. In that way the users can choose how to leverage dynamic analysis as they wish

spoiicy replied:
Sure I'll add them as parameters for users to pass a desired value.

mlodic reviewed (changes requested):
thanks for the PR!

spoiicy commented:
Hi @mlodic I've made the changes as per your comments. I hope we are good to go, let me know otherwise. :)

mlodic commented on api_app/analyzers_manager/migrations/0141_analyzer_config_mobsf_service.py:
```diff
@@ -0,0 +1,219 @@
+    "name": "MobSF_Service",
+    "description": "[MobSF_service](https://github.com/MobSF/Mobile-Security-Framework-MobSF) can be used for a variety of use cases such as mobile application security, penetration testing, malware analysis, and privacy analysis.",
+    "disabled": False,
+    "soft_time_limit": 400,
```
as mentioned in a previous comment, you can extend it to 1000 to allow for broader use cases

mlodic reviewed (changes requested):
I have to ask you other stuff to have this PR accepted. Some standards needs to be followed:

* a PR against the documentation repository has been added? Please do that following the checklist and add a link to that PR here.
* the static_analysis and dynamic_analysis functions are really difficult to read and they contain a lot or redundant code. I ask you to refactor that code in a way that there is an helper method that executes all of the requests stuff

spoiicy commented:
@mlodic I've added the usage for MobSF_Service. [Here](https://github.com/intelowlproject/docs/pull/18) is the pull request.
I've improved the readability of the code by adding helper method to handle request-response and refactored some of the code. 

The PR is ready to be reviewed.

mlodic commented:
thank you, we are ready to merge it!
